src/devices_models/devices/common/add_variable.jl

```diff
@@ -1,3 +1,134 @@
+struct AddVariableInputs
+    variable_names::Vector{Symbol}
+    binary::Bool
+    expression_name::Union{Nothing, Symbol}
+    sign::Float64
+    devices_filter_func::Union{Nothing, Function}
+    initial_value_func::Union{Nothing, Function}
+    lb_value_func::Union{Nothing, Function}
+    ub_value_func::Union{Nothing, Function}
+end
+
+"""
+Construct AddVariableInputs.
+
+Accepts a single variable_name or a vector variable_names. One must be passed but not both.
+"""
+function AddVariableInputs(;
+    variable_name = nothing,
+    variable_names = nothing,
+    binary,
+    expression_name = nothing,
+    sign = 1.0,
+    devices_filter_func = nothing,
+    initial_value_func = nothing,
+    lb_value_func = nothing,
+    ub_value_func = nothing,
+)
+    if isnothing(variable_name) && isnothing(variable_names)
+        throw(ArgumentError("either variable_name or variable_names must be set"))
+    end
+    if !isnothing(variable_name) && !isnothing(variable_names)
+        throw(ArgumentError("variable_name and variable_names cannot both be set"))
+    end
+    if !isnothing(variable_name)
+        variable_names = [variable_name]
+    end
+
+    return AddVariableInputs(
+        variable_names,
+        binary,
+        expression_name,
+        sign,
+        devices_filter_func,
+        initial_value_func,
+        lb_value_func,
+        ub_value_func,
+    )
+end
+
+function make_variable_inputs(
+    ::Type{<:T},
+    ::Type{<:U},
+    ::PSIContainer,
+) where {T <: VariableType, U <: PSY.Component}
+    error("make_variable_inputs is not implemented for $T / $U")
+end
+
+"""
+Add variables to the PSIContainer for any component.
+"""
+function add_variables!(
+    ::Type{T},
+    psi_container::PSIContainer,
+    devices::IS.FlattenIteratorWrapper{U},
+) where {T <: VariableType, U <: PSY.Component}
+    _add_variables!(psi_container, devices, make_variable_inputs(T, U, psi_container))
+end
+
+"""
+Add variables to the PSIContainer for a service.
+"""
+function add_variables!(
+    ::Type{T},
+    psi_container::PSIContainer,
+    service::U,
+    contributing_devices::Vector{V},
+) where {T <: VariableType, U <: PSY.Reserve, V <: PSY.Device}
+    inputs = make_variable_inputs(T, psi_container, service, V)
```

Comment:
There may be some disagreement here.  I am passing the type of the contributing devices, not the devices.  The filter function gets assigned in this function and applied in the next.

Reply:
This won't work generally since contributing devices aren't from a single ConcreteType. For instance, reserves can be provided by ThermalStandard and HydroDispatch.

Reply:
The type of each contributing device isn't actually used until `add_variable` is called at the end of the sequence, so I don't think it matters.  It is only used for parameterization, so we could remove it altogether.  Unless I'm missing something...

Reply:
The issue is that we don't know what V will be. We wouldn't in principle even need to parametrize here.

Reply:
Then I'll remove it.